vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py

```diff
@@ -412,75 +425,94 @@ def __init__(self, vllm_config: VllmConfig, engine_id: str):
-                metadata_bytes = sock.recv()
-                decoder = msgspec.msgpack.Decoder(NixlAgentMetadata)
-                metadata = decoder.decode(metadata_bytes)
-                got_metadata_time = time.perf_counter()
-
-                # Register Remote agent.
-                self.add_remote_agent(metadata, rank)
-                setup_agent_time = time.perf_counter()
-
-                logger.debug("NIXL handshake: get metadata took: %s",
-                             got_metadata_time - start_time)
-                logger.debug("NIXL handshake: add agent took: %s",
-                             setup_agent_time - got_metadata_time)
-                return metadata
-
-        # Handshake with remote agent-rank0 first to get the tp_size of remote
-        path = make_zmq_path("tcp", host, port)
-        logger.debug("Querying master rank metadata on path: %s", path)
-        metadata = handshake(path, 0)
+        # TODO: make the scheme dynamic, and/or implement https on both sides.
+        url = build_uri("http", host, port, path="get_kv_connector_metadata")
+        logger.debug("Querying metadata on path: %s", url)
+
+        try:
+            req = Request(url)
+            with urlopen(req, timeout=5.0) as response:
+                response_data = response.read().decode('utf-8')
+                res = json.loads(response_data)
+                logger.debug("NIXL handshake response: %s", res)
+        except (URLError, HTTPError) as e:
+            logger.error("Failed to fetch metadata from %s: %s", url, e)
+            raise
+
+
+        remote_tp_size = len(res.keys())
+        # Default case is that the remote TP size is 1, so we can
+        # directly access the metadata.
+        tp_data = res.get(str(self.tp_rank), {}).get("0", {})
+        metadata_bytes = tp_data.get("agent_metadata", None)
 
         # Handshake only with the other TP remote the current local rank will
         # pull from. With homogeneous TP it happens to be the same rank_i.
-        tp_ratio = self._tp_size[self.engine_id] // metadata.tp_size
+        tp_ratio = self._tp_size[self.engine_id] // remote_tp_size
         p_remote_rank = self.tp_rank // tp_ratio
         if p_remote_rank > 0:
-            path = make_zmq_path("tcp", host, port + p_remote_rank)
-            logger.debug("Querying metadata on path: %s at remote rank %s",
-                         path, p_remote_rank)
-            _ = handshake(path, p_remote_rank)
+            metadata_bytes = res.get(str(p_remote_rank), {}).get("0", {}).get(
+                "agent_metadata", None)
+
+        if metadata_bytes is not None:
+            # Reconstruct NixlAgentMetadata from JSON response
+            # agent_metadata is base64-encoded binary data, not msgpack
+            metadata = NixlAgentMetadata(
+                engine_id=tp_data["engine_id"],
+                agent_metadata=base64.b64decode(metadata_bytes),
+                kv_caches_base_addr=tp_data["kv_caches_base_addr"],
```

Comment:
ok I see here the problem with sending a lot of memory pointers through json..a bit ugly, but it's a one off

Reply:
yeah... json might not be ideal here since the agent_metadata is binary. we can revisit this encoding, since it's an internal API we could probably just use msgpack.